### PR TITLE
Do not use Count() when Any() can be used.

### PR DESCRIPTION
**Pull request by paulomorgado**
# Do not use Count() when Any() can be used.
`System.Linq.Enumerable.Count<TSource>(System.Collections.Generic.IEnumerable<TSource>)"` enumerates the entire enumerable while `System.Linq.Enumerable.Any<TSource>(System.Collections.Generic.IEnumerable<TSource>)` will only enumerates, at most, up until the first item.

## Use cases

detected | fix
-------- | --- 
`enumerable.Count() == 0` | `!enumerable.Any()`
`enumerable.Count() != 0` | `enumerable.Any()`
`enumerable.Count() <= 0` | `!enumerable.Any()`
`enumerable.Count() > 0` | `enumerable.Any()`
`enumerable.Count() < 1` | `!enumerable.Any()`
`enumerable.Count() >= 1` | `enumerable.Any()`
`0 == enumerable.Count()` | `!enumerable.Any()`
`0 != enumerable.Count()` | `enumerable.Any()`
`0 >= enumerable.Count()` | `!enumerable.Any()`
`0 < enumerable.Count()` | `enumerable.Any()`
`1 > enumerable.Count()` | `!enumerable.Any()`
`1 <= enumerable.Count()` | `enumerable.Any()`
`enumerable.Count().Equals(0)` | `!enumerable.Any()`
`0.Equals(enumerable.Count())` | `!enumerable.Any()`
`enumerable.Count(_ => true) == 0` | `!enumerable.Any(_ => true)`
`enumerable.Count(_ => true) != 0` | `enumerable.Any(_ => true)`
`enumerable.Count(_ => true) <= 0` | `!enumerable.Any(_ => true)`
`enumerable.Count(_ => true) > 0` | `enumerable.Any(_ => true)`
`enumerable.Count(_ => true) < 1` | `!enumerable.Any(_ => true)`
`enumerable.Count(_ => true) >= 1` | `enumerable.Any(_ => true)`
`0 == enumerable.Count(_ => true)` | `!enumerable.Any(_ => true)`
`0 != enumerable.Count(_ => true)` | `enumerable.Any(_ => true)`
`0 < enumerable.Count(_ => true)` | `!enumerable.Any(_ => true)`
`0 >= enumerable.Count(_ => true)` | `enumerable.Any(_ => true)`
`1 > enumerable.Count(_ => true)` | `!enumerable.Any(_ => true)`
`1 <= enumerable.Count(_ => true)` | `enumerable.Any(_ => true)`
`enumerable.Count(_ => true).Equals(0)` | `!enumerable.Any(_ => true)`
`0.Equals(enumerable.Count(_ => true))` | `!enumerable.Any(_ => true)`

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/Microsoft.CodeQuality.Analyzers.csproj [hunk 1]
```diff
@@ -14,6 +14,11 @@
     <InternalsVisibleTo Include="Microsoft.CodeQuality.Analyzers.UnitTests" />
     <PackageReference Include="Humanizer.Core" Version="$(HumanizerVersion)" />
   </ItemGroup>
+  <ItemGroup>
+    <EmbeddedResource Update="Performance\MicrosoftPerformanceAnalyzersResources.resx">
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Microsoft.CodeQuality.Analyzers.csproj]: Please revert this change to project file. I believe the SDK already takes care of this...

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx [hunk 2]
```diff
@@ -0,0 +1,129 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx]: Can we avoid creating new resource files and folder(s) for this rule here? You can move the rule under "QualityGuidelines" folder and use the existing resx file. For example, see this performance rule under it: https://github.com/dotnet/roslyn-analyzers/blob/ac2cae9dd4807007e7ae0d25884a3d734703636d/src/Microsoft.CodeQuality.Analyzers/Core/QualityGuidelines/PreferJaggedArraysOverMultidimensional.cs#L32

FYI: Multiple resx files inside a single project is an unfortunate consequence of us previously having these rules spread across different projects such as Maintenance, QualityGuidelines, etc. That was later refactored to have all of them in a  single Microsoft.CodeQuality.Analyzers project, and we just moved the sources and resx files as is. We haven't yet got to merging all the resx files into a single one, but we should certainly avoid creating more resx files inside each project.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        /// <summary>
+        /// Called on compilation start.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var enumerableType = context.Compilation.GetTypeByMetadataName(EnumerableMetadataName);
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Please add to https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/WellKnownTypes.cs and https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/WellKnownTypeNames.cs

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx (hunk 2 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx]: @mavasani,

I followed what I perceived to be the standard for the projects.

Are saying to move the analyzer from the `Performance` folder to the `QualityGuidelines` folder or to use the resources in the `QualityGuidelines` folder? I don't think either one makes sense because it would break the relation between the namespace and the folder name.

Wouldn't it make more sense to have a top level resource file for each project and move all resources there?

**Comment by paulomorgado** [comment]
Added fixer!

File: src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: Usings before copyright

File: RoslynAnalyzers.sln.GhostDoc.xml [hunk 5]
```diff
@@ -0,0 +1,60 @@
+<GhostDoc>
```
Comment (mavasani) [inline RoslynAnalyzers.sln.GhostDoc.xml]: Want to revert?

File: src/Microsoft.CodeAnalysis.FxCopAnalyzers/Microsoft.CodeAnalysis.FxCopAnalyzers.sarif [hunk 6]
```diff
@@ -1258,6 +1258,26 @@
...
+          "properties": {
+            "category": "Performance",
+            "isEnabledByDefault": true,
+            "typeName": "DoNotUseCountWhenAnyCanBeUsedAnalyzer",
+            "languages": [
+              "C#",
+              "Visual Basic"
+            ],
+            "tags": [
+              "PortedFxCopRuleTag",
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.FxCopAnalyzers/Microsoft.CodeAnalysis.FxCopAnalyzers.sarif]: This should not be a ported FxCop rule

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx]: Sorry, I meant what you suggested in the last statement is _exactly_ what we should do - have a single resource file per project. We have multiple of them because of historical reasons, and I'd prefer if we not add one more to the project.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+        /// <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource})"/> or
+        /// <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource}, Func{TSource, bool})"/>
+        /// using <see cref="int.Equals(int)"/>.
+        /// </summary>
+        private static void AnalyzeInvocationExpression(IInvocationOperation invocationOperation, INamedTypeSymbol enumerableType, Action<Diagnostic> reportDiagnostic)
+        {
+            if (invocationOperation.Arguments.Length == 1)
+            {
+                var methodSymbol = invocationOperation.TargetMethod;
+                if (methodSymbol != null &&
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: TargetMethod is guaranteed to be non-null, you can remove the null check here.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx (hunk 2 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/MicrosoftPerformanceAnalyzersResources.resx]: Moved just the resources for this analyzer.
I'll come up with something to batch change all others.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+
+        /// <summary>
+        /// Check to see if we have an expression comparing the result of
+        /// <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource})"/> or
+        /// <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource}, Func{TSource, bool})"/>
+        /// using operators.
+        /// </summary>
+        private static void AnalyzeBinaryExpression(IBinaryOperation binaryOperation, INamedTypeSymbol enumerableType, Action<Diagnostic> reportDiagnostic)
+        {
+            if (binaryOperation.OperatorKind != BinaryOperatorKind.Equals &&
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Nit: Might be cleaner to move it to an extension method, say `AnyBinaryEqualsOperatorKind`, and use a switch case.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+        /// <param name="operation">The operation.</param>
+        /// <param name="constant">The constant the <paramref name="operation"/> represents, if succeeded, or zero if <paramref name="operation"/> is not a constant.</param>
+        /// <returns><see langword="true" /> <paramref name="operation"/> is a constant, <see langword="false" /> otherwise.</returns>
+        public static bool TryGetInt32Constant(IOperation operation, out int constant)
+        {
+            constant = default;
+
+            if (operation.Type?.SpecialType != SpecialType.System_Int32)
+            {
+                // TODO: Should we be checking for other integer types?
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Don't think we need to, probably remove the TODO?

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+        /// <summary>
+        /// Checks the is a <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource})" />
+        /// or <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource}, Func{TSource, bool})" /> invocation.
+        /// </summary>
+        /// <param name="operation">The operation.</param>
+        /// <param name="enumerableType">Type of the enumerable.</param>
+        /// <returns><see langword="true" /> if XXXX, <see langword="false" /> otherwise.</returns>
+        private static bool IsCountMethodInvocation(IOperation operation, INamedTypeSymbol enumerableType)
+        {
+            if (operation.Kind != global::Microsoft.CodeAnalysis.OperationKind.Invocation ||
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Any reason to use fully qualified names instead of just having a `using Microsoft.CodeAnalysis` and `using Microsoft.CodeAnalysis.Operations`?

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+                !(operation is global::Microsoft.CodeAnalysis.Operations.IInvocationOperation invocationExpression) ||
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Kind check above is redundant as you are already doing a pattern match check.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+                !invocationExpression.TargetMethod.ContainingSymbol.Equals(enumerableType) ||
+                !invocationExpression.TargetMethod.Name.Equals(DoNotUseCountWhenAnyCanBeUsedAnalyzer.CountMethodName, StringComparison.Ordinal))
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Move the cheaper name check ahead of the containing symbol equality check.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+            {
+                return false;
+            }
+
+            // TODO: Should I be doing more checks?
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Doesn't seem more checks are required.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+
+            var parameterCountOFfset = operation.Language == LanguageNames.VisualBasic ? -1 : 0;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: We normally don't have such language specific checks in language agnostic layer. Also can you add a comment and explain why this is needed.

Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Looking at the code below, personally I don't think you need any of it - if a method named `Count` is invoked on `enumerableType`, we can assume it is for count computation regardless of what parameters it has. We do this as almost all places where well known method invocation checks are performed.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+        /// using <see cref="int.Equals(int)"/>.
+        /// </summary>
+        /// <param name="invocationOperation">The invocation operation.</param>
+        /// <param name="enumerableType">Type of the enumerable.</param>
+        /// <returns><see langword="true" /> if 0 is being compared with the value of the invocation of 
+        /// <see cref="System.Linq.Enumerable.Count{TSource}(System.Collections.Generic.IEnumerable{TSource})"/>
+        /// using <see cref="int.Equals(int)"/>; otherwise, <see langword="false" />.</returns>
+        private static bool IsZeroEqualsCount(IInvocationOperation invocationOperation, INamedTypeSymbol enumerableType)
+        {
+            if (!TryGetInt32Constant(invocationOperation.Instance, out var constant) || constant != 0)
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Can `Instance` be null for error cases?

File: src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="DoNotUseCountWhenAnyCanBeUsedDescription" xml:space="preserve">
+    <value>Count() enumerates the entire enumerable while Any() will only enumerates, at most, up until the first item.</value>
+  </data>
+  <data name="DoNotUseCountWhenAnyCanBeUsedMessage" xml:space="preserve">
+    <value>Count() is being used where Any() could be used.</value>
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: @gewarren for double checking the diagnostic message.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        /// <param name="context">A <see cref="CodeFixContext" /> containing context information about the diagnostics to fix.
+        /// The context must only contain diagnostics with a <see cref="Diagnostic.Id" /> included in the <see cref="CodeFixProvider.FixableDiagnosticIds" /> 
+        /// for the current provider.</param>
+        /// <returns>A <see cref="Task" /> that represents the asynchronous operation.</returns>
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span);
+
+            if (this.TryGetFixer(node, out var expression, out var arguments, out var negate))
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: Probably want to have a defensive bail out check for `node == null` before invoking this method.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                var title = MicrosoftCodeQualityAnalyzersResources.DoNotUseCountWhenAnyCanBeUsedTitle;
+                context.RegisterCodeFix(
+                    CodeAction.Create(
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: We use a privately defined code action type, such as https://github.com/dotnet/roslyn-analyzers/blob/32d8f1e397439035f0ecb5f61a9e672225f0ecdb/src/Microsoft.NetCore.Analyzers/Core/Runtime/AvoidUnsealedAttributes.Fixer.cs#L51, primarily for telemetry purposes. See https://github.com/dotnet/roslyn/issues/4919 for details.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs [hunk 19]
```diff
@@ -0,0 +1,550 @@
...
+");
+        }
+
+        #endregion
+
+        #region Unit tests for analyzer diagnostic(s)
+
+        [Theory]
+        [MemberData(nameof(DoNotUseCountWhenAnyCanBeUsedTestData.LeftCount_Diagnostic_TheoryData), MemberType = typeof(DoNotUseCountWhenAnyCanBeUsedTestData))]
+        public void CSharp_Diagnostic_LeftCount(BinaryOperatorKind @operator, int value)
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs]: 💡  _All_ diagnostic tests should use `VerifyCodeFixAsync`. Separating the code fix tests serves only to increase complexity and increase the probability of a test gap. I'm slowly working to update the old tests in this repository to use the recommended pattern, but new tests for analyzers that have a fixer should start that way.

File: src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+                GetFixerForGreaterThanExpression((BinaryExpressionSyntax)node, out expression, out arguments, out negate);
+                return true;
+            }
+            else if (node.IsKind(SyntaxKind.GreaterThanOrEqualExpression))
+            {
+                GetFixerForGreaterThanOrEqualExpression((BinaryExpressionSyntax)node, out expression, out arguments, out negate);
+                return true;
+            }
+
+            expression = default;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: Is this code path possible? I understand if this is just a defensive approach...

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs [hunk 21]
```diff
@@ -0,0 +1,550 @@
...
+using Microsoft.CodeQuality.CSharp.Analyzers.Maintainability;
+using Microsoft.CodeQuality.VisualBasic.Analyzers.Maintainability;
+using Test.Utilities;
+using Xunit;
+using static Microsoft.CodeQuality.Analyzers.Performance.UnitTests.DoNotUseCountWhenAnyCanBeUsedTestData;
+
+namespace Microsoft.CodeQuality.Analyzers.Performance.UnitTests
+{
+
+    public class DoNotUseCountWhenAnyCanBeUsedAnalyzerTests : DiagnosticAnalyzerTestBase
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs]: 💡 `DiagnosticAnalyzerTestBase` is the legacy testing harness. New tests should use `VerifyCS` and `VerifyVB`.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Fixer.cs [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.Performance.DoNotUseCountWhenAnyCanBeUsedAnalyzer,
+    Microsoft.CodeQuality.CSharp.Analyzers.Performance.CSharpDoNotUseCountWhenAnyCanBeUsedFixer>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.CodeQuality.Analyzers.Performance.DoNotUseCountWhenAnyCanBeUsedAnalyzer,
+    Microsoft.CodeQuality.VisualBasic.Analyzers.Performance.BasicDoNotUseCountWhenAnyCanBeUsedFixer>;
+
+namespace Microsoft.CodeQuality.Analyzers.Performance.UnitTests
+{
+    public static class DoNotUseCountWhenAnyCanBeUsedFixerTests
```
Comment (sharwell) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.Fixer.cs]: 💡 Code fix tests do not need to be separated into their own file. All analyzer tests using the new test harness are also code fix tests, so there is no real distinguishing between the two.

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs (hunk 17 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: Any reason why the `TryGetFixer` invocation is done before registering a code action? As in will that be realistic cases where this method returns false? Otherwise, we can always register a code action and move this invocation inside FixAsync.

If you feel the current approach is safer, then totally fine to keep your current implementation.

**Comment by mavasani** [comment]
Seems like you have valid build failure due to CA1801 diagnostics.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs (hunk 19 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs]: @sharwell, do you mean that, in this particular case, I should remove these tests and keep only the fixer tests. They are the same use cases.

File: src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs (hunk 20 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/CSharp/Performance/CSharpDoNotUseCountWhenAnyCanBeUsed.Fixer.cs]: It's defensive. Just in case. 😄

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs (hunk 19 above)
Comment (sharwell) [reply src/Microsoft.CodeQuality.Analyzers/UnitTests/Performance/DoNotUseCountWhenAnyCanBeUsedTests.cs]: @paulomorgado If you have a diagnostic test and a code fix test that have the same input source code, then you can simply remove the diagnostic test and keep the existing code fix test. `VerifyCodeFixAsync` automatically verifies the diagnostics of both the source and fixed code, so the code fix in a case like this would be exactly duplicating work done in the diagnostic test.

However, it's common for users getting started with the new test framework to find that one or more diagnostic tests do not already have a code fix test present with the same input. In those cases, it would be good to update the diagnostic test rather than simply removing it.

File: RoslynAnalyzers.sln.GhostDoc.xml (hunk 5 above)
Comment (mavasani) [reply RoslynAnalyzers.sln.GhostDoc.xml]: @paulomorgado Seems like it came back again :)

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs [hunk 23]
```diff
@@ -0,0 +1,315 @@
...
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.DoNotUseCountWhenAnyCanBeUsedDescription), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources));
+        private static readonly DiagnosticDescriptor s_rule = new DiagnosticDescriptor(
+            RuleId,
+            Title,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultForVsixAndNuget,
+            description: s_localizableDescription,
+            helpLinkUri: "https://docs.microsoft.com/visualstudio/code-quality/" + RuleId);
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Do we need this will be the help link URI? Otherwise, we can let it be null and add one later. Tagging @gewarren

File: src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    <value>2.0</value>
+  </resheader>
+  <resheader name="reader">
+    <value>System.Resources.ResXResourceReader, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <resheader name="writer">
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="DoNotUseCountWhenAnyCanBeUsedDescription" xml:space="preserve">
+    <value>Count() enumerates the entire enumerable while Any() will only enumerates, at most, up until the first item.</value>
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: > will only enumerates, at most, up until the first item

Typo? Should this be `will only enumerate one item`? @gewarren

File: src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs (hunk 23 above)
Comment (dotpaul) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: Nit: I think docs.microsoft.com URLs are usually all lowercase (the website seems to redirect from uppercase to lowercase, but still).

Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs]: I'll lower case it. I'm just trying to avoid the need to change the analyzer to add a predictable text.

File: src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx (hunk 24 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: If there are no items, it won't enumerate any.

Comment (drewnoakes) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: @mavasani I agree with you. As it stands the sentence reads as incorrect.

Perhaps use:

> Count() enumerates the entire sequence, while Any() stops at the first item.

Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: > If the sequence has items, Count() enumerates the entire sequence, while Any() stops at the first item.

Comment (gewarren) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: Yes as written the grammar is incorrect. One question - are we only referring to the overload of Any() that doesn't take any parameters, or must we also consider the overload that looks for an item that satisfies a condition? That might change the wording a bit. If it's just the no-parameter overload, I'd suggest something like:

For performance reasons, use Any(), which enumerates at most one item, instead of Count(), which enumerates the entire sequence.

Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: Both!

`Count()` will be converted to `Any()` and `Cont(predicate)` will be converted to `Any(predicate)`.

Comment (gewarren) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: I see now that this is the description and not the message, so perhaps a variation of drewnoakes' suggestion such as:

Count() enumerates the entire sequence, while Any() stops at the first item or the first item that satisfies a condition.

File: src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx (hunk 16 above)
Comment (gewarren) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]:
```suggestion
    <value>Count() is used where Any() could be used instead to improve performance.</value>
```

File: src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx (hunk 24 above)
Comment (paulomorgado) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: And what about empty collections? Any doesn't stop at the first item because there isn't one.

Comment (gewarren) [reply src/Microsoft.CodeQuality.Analyzers/Core/MicrosoftCodeQualityAnalyzersResources.resx]: How about:

For non-empty collections, Count() enumerates the entire sequence, while Any() stops at the first item or the first item that satisfies a condition.
